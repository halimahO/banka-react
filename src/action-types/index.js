export const LOGIN_SUCCESS = 'LOGIN_SUCCESS';
export const LOGIN_ERROR = 'LOGIN_ERROR';
export const SIGNUP_SUCCESS = 'SIGNUP_SUCCESS';
export const SIGNUP_ERROR = 'SIGNUP_ERROR';
export const CREATE_ACCOUNT_SUCCESS = 'CREATE_ACCOUNT_SUCCESS';
export const CREATE_ACCOUNT_ERROR = 'CREATE_ACCOUNT_ERROR';
export const SET_CURRENT_USER = 'SET_CURRENT_USER';
export const USER_TRANSACTION_SUCCESS = 'USER_TRANSACTION_SUCCESS';
export const USER_TRANSACTION_ERROR = 'USER_TRANSACTION_ERROR';
export const DEBIT_SUCCESS = 'DEBIT_SUCCESS';
export const DEBIT_ERROR = 'DEBIT_ERROR';
export const CREDIT_SUCCESS = 'CREDIT_SUCCESS';
export const CREDIT_ERROR = 'CREDIT_ERROR';
