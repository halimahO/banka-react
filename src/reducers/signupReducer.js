import * as types from '../action-types/index';

const signupReducer = (state = {}, action) => {
  switch (action.type) {
    case types.SIGNUP_ERROR:
      return { ...state, error: action.payload };
    case types.SIGNUP_SUCCESS:
      return { ...state, user: action.payload };
    default:
      return state;
  }
};

export default signupReducer;
