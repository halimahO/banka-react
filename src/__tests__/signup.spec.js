import moxios from 'moxios';
import axios from 'axios';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { toast } from 'react-toastify';
import { signupAction, signupError } from '../actions/signup.action';
import { SIGNUP_SUCCESS, SIGNUP_ERROR } from '../action-types/index';
import signupReducer from '../reducers/signupReducer';

jest.mock('axios');

describe('Signup Actions', () => {
  const userCredentials = {
    firstName: 'Abu',
    lastName: 'Isaac',
    email: 'someuser@gmail.com',
    password: 'randompassword'
  };

  const middleware = [thunk];
  const mockStore = configureMockStore(middleware);
  let store;
  beforeEach(() => {
    store = mockStore({});
    jest.resetAllMocks();
  });
  afterEach(() => {
    store.clearActions();
  });

  it('Should Trigger the SIGN_UP_USER dispatch function', async () => {
    jest.useFakeTimers();
    const mockData = {
      status: 201,
      data: {
        data: {
          token: 'eyJhbGlsIIDrYp9urUBU7UBvM9LUBn_Vachtw',
          id: 1,
          firstname: 'Ade',
          lastname: 'Bade',
          email: 'adebae@gmail.com',
          type: 'staff'
        }
      }
    };

    axios.post.mockResolvedValueOnce(mockData);

    const expectedActions = [{ type: 'SIGNUP_SUCCESS', payload: mockData }];
    const historyObject = {
      push: jest.fn()
    };

    store.dispatch(signupAction(userCredentials, historyObject)).then(() => {
      jest.runOnlyPendingTimers();
      expect(store.getActions()).toEqual(expectedActions);
    });
  });

  it('Should Trigger the SIGNUP_ERROR dispatch function', async () => {
    const mockData = {
      response: {
        data: {
          status: 409,
          error: 'This email address is already taken.'
        }
      }
    };

    axios.post.mockRejectedValueOnce(mockData);

    const historyObject = {
      push: jest.fn()
    };
    const expectedActions = [
      { type: 'SIGNUP_ERROR', payload: mockData.response.data.error }
    ];
    store.dispatch(signupAction(userCredentials, historyObject)).then(() => {
      expect(store.getActions()).toEqual(expectedActions);
    });
  });

  it('Should return a toast on network error', async () => {
    const mockData = {
      message: 'Network Error'
    };

    const spy = jest.spyOn(toast, 'error');

    axios.post.mockRejectedValueOnce(mockData);

    store.dispatch(signupAction(userCredentials, null)).then(() => {
      expect(spy).toHaveBeenCalledWith('An error occured. Please try Again');
    });
  });
});

describe('Signup Reducer', () => {
  it('Should return a new state if it recieves a signup success action type', () => {
    const user = {
      id: 97,
      firstname: 'Lola',
      lastname: 'Lara',
      email: 'lara@mail.com',
      type: 'savings',
      token: 'sometoken'
    };
    const newState = signupReducer(undefined, {
      type: SIGNUP_SUCCESS,
      payload: user
    });
    expect(newState).toEqual({
      user
    });
  });

  it('Should return a new state if it recieves signup error action type', () => {
    const message = 'invalid username/password';
    const newState = signupReducer(undefined, {
      type: SIGNUP_ERROR,
      payload: message
    });
    expect(newState).toEqual({
      error: message
    });
  });
});
