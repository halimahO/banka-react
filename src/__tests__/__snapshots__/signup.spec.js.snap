// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup component Tests renders the Login component correctly 1`] = `ShallowWrapper {}`;
